man/git-ftp.1.md
% GIT-FTP(1) Git-ftp 1.6.0
%
% 2020-02-03

This is the manual for version 1.6.0-UNRELEASED.
Please consider the [changelog](../CHANGELOG.md) or select your version from
the _Branch > Tags_ select above to see the manual for another version.

# NAME

Git-ftp - Git powered FTP client written as shell script.

# SYNOPSIS

*git-ftp* \<action\> [\<options\>] [\<url\>]

# DESCRIPTION

Git-ftp is an FTP client using [Git] to determine which local files to upload or
which files to delete on the remote host.

It saves the deployed state by uploading the SHA1 hash in the `.git-ftp.log`
file. There is no need for Git to be installed on the remote host.

Even if you play with different branches, git-ftp knows which files are
different and handles only those files. That saves time and bandwidth.

# ACTIONS

`init`
:	Uploads all git-tracked non-ignored files to the remote server and
	creates the `.git-ftp.log` file containing the SHA1 of the latest
	commit.

`catchup`
:	Creates or updates the `.git-ftp.log` file on the remote host.
	It assumes that you uploaded all other files already.
	You might have done that with another program.

`push`
:	Uploads files that have changed and
	deletes files that have been deleted since the last upload.
	If you are using GIT LFS, this uploads LFS link files, 
	not large files (stored on LFS server). 
	To upload the LFS tracked files, run `git lfs pull`
	before `git ftp push`: LFS link files will be replaced with 
	large files so they can be uploaded.  

`download` (EXPERIMENTAL)
:	Downloads changes from the remote host into your working tree.
	This feature needs lftp to be installed and does not use any power of
	Git.
	WARNING: It can delete local untracked files that are not listed in
	your `.git-ftp-ignore` file.

`pull` (EXPERIMENTAL)
:	Downloads changes from the remote host into a separate commit
	and merges that into your current branch.
	If you just want to download the files without a merge, consider `download`.
	This feature needs lftp to be installed.

`snapshot` (EXPERIMENTAL)
:	Downloads files into a new Git repository. Takes an additional
	argument as local destination directory. Example:
	\`git-ftp snapshot ftp://example.com/public_html projects/example\`
	This feature needs lftp to be installed.

`show`
:	Downloads last uploaded SHA1 from log and hooks \`git show\`.

`log`
:	Downloads last uploaded SHA1 from log and hooks \`git log\`.

`add-scope <scope>`
:	Creates a new scope (e.g. dev, production, testing, foobar).
	This is a wrapper action over git-config.
	See **SCOPES** section for more information.

`remove-scope <scope>`
:	Remove a scope.

`help`
:	Shows a help screen.

# OPTIONS

`-u [username]`, `--user [username]`
:	FTP login name. If no argument is given, local user will be taken.

`-p [password]`, `--passwd [password]`
:	FTP password. See `-P` for interactive password prompt. ([note](#passwords))

`-P`, `--ask-passwd`
:	Ask for FTP password interactively.

`--password-command [command]`
:	FTP password from arbitrary bash command.

`-k [[account]@[host]]`, `--keychain [[account]@[host]]`
:	FTP password from KeyChain (macOS only).

`-a`, `--all`
:	Uploads all files of current Git checkout.

`-c`, `--commit`
:	Sets SHA1 hash of last deployed commit by option.

`-A`, `--active`
:	Uses FTP active mode. This works only if you have either no firewall
	and a direct connection to the server or an FTP aware firewall. If you
	don't know what it means, you probably won't need it.

`-b [branch]`, `--branch [branch]`
:	Push a specific branch

`-s [scope]`, `--scope [scope]`
:	Using a scope (e.g. dev, production, testing, foobar). See **SCOPE**
	and **DEFAULTS** section for more information.

`-l`, `--lock`
:	Enable remote locking.

`-D`, `--dry-run`
:	Does not upload or delete anything, but tries to get the `.git-ftp.log`
	file from remote host.

`-f`, `--force`
:	Does not ask any questions, it just does.

`-n`, `--silent`
:	Be silent.

`-h`, `--help`
:	Prints some usage information.

`-v`, `--verbose`
:	Be verbose.

`-vv`
:	Be as verbose as possible. Useful for debug information.

`--remote-root`
:	Specifies the remote root directory to deploy to.
	The remote path in the URL is ignored.

`--syncroot`
:	Specifies a local directory to sync from as if it were the git project
	root path.

`--key`
:	SSH private key file name for SFTP.

`--pubkey`
:	SSH public key file name. Used with --key option.

`--insecure`
:	Don't verify server's certificate.

`--cacert <file>`
:	Use <file> as CA certificate store.
	Useful when a server has a self-signed certificate.

`--disable-epsv`
:	Tell curl to disable the use of the EPSV command when doing passive FTP
	transfers.
	Curl will normally always first attempt to use EPSV before PASV,
	but with this option, it will not try using EPSV.

`--no-commit`
:	Stop while merging downloaded changes during the pull action.
	A commit is made anyway, but the merge is interrupted.
	If you just want to download the files you could also consider the action
	`download`.

`--changed-only`
:	During the ftp mirror operation during a pull command, consider only
	the files changed since the deployed commit.

`--no-verify`
:	Bypass the pre-ftp-push hook. See **HOOKS** section.

`--no-post-hooks`
:	Bypass the post-ftp-push hook. See **HOOKS** section.

`--enable-post-errors`
:	Fails if post-ftp-push raises an error.

`--auto-init`
:	Automatically run init action when running push action

`--version`
:	Prints version.

`-x [protocol://]host[:port]`, `--proxy [protocol://]host[:port]`
:	Use the specified proxy. This option is passed to curl.
	See the curl manual for more information.

# URL

The scheme of an URL is what you would expect

	protocol://host.domain.tld:port/path

Below a full featured URL to *host.example.com* on port *2121* to path *mypath*
using protocol *ftp*:

	ftp://host.example.com:2121/mypath

But, there is not just FTP. Supported protocols are:

`ftp://...`
:	FTP (default if no protocol is set)

`sftp://...`
:	SFTP

`ftps://...`
:	FTPS

`ftpes://...`
:	FTP over explicit SSL (FTPES) protocol

# EXAMPLES

## FIRST UPLOADS

Upload your files to an FTP server the first time:

	$ git ftp init -u "john" -P "ftp://example.com/public_html"

It will authenticate with the username `john` and ask for the
password. By default, it tries to transfer data in EPSV mode.
Depending on the network and server configuration, that may fail.
You can try to add the `--disable-epsv` option to use the IPv4 passive FTP
connection (PASV). In rare circumstances, you can use `--active` for the
original FTP transfer mode. These options do not apply to SFTP.

You are less likely to face connection problems with SFTP.
But be aware of the different
handling of relative and absolute paths. If the directory `public_html` is in
the home directory on the server, then upload like this:

	$ git ftp init -u "john" --key "$HOME/.ssh/id_rsa" "sftp://example.com/~/public_html"

Otherwise it will use an absolute path, for example:

	$ git ftp init -u "john" --key "$HOME/.ssh/id_rsa" "sftp://example.com/var/www"

On some systems Git-ftp fails to verify the server's fingerprint.
You can then use the `--insecure` option to skip the verification.
That will leave you vulnerable to man-in-the-middle attacks, but is still more
secure than plain FTP.

Git-ftp guesses the path of the public key file corresponding to your private
key file. If you just have a private key, for example a .pem file, you need
Git-ftp version 1.3.4 and Curl version 7.39.0 or newer.
If you have an older version of Git-ftp or Curl, you can
create the public key with the ssh-keygen command:

	$ ssh-keygen -y -f key.pem > key.pem.pub

## RESET THE UPLOADED STATE

Many people already uploaded their files to the server.
If you want to mark the uploaded version as the same as your local branch:

	$ git ftp catchup

This example omits options like `--user`, `--password` and `url`.
See DEFAULTS below to learn how to store your configuration so that you don't
need to repeat it.

After you stored the commit id of the uploaded commit via `init` or
`catchup`, you can then upload any new commits:

	$ git ftp push

If you discovered a bug in the last uploaded version and you want to go back
by three commits:

	$ git checkout HEAD~3
	$ git ftp push

Or maybe some files got changed on the server and you want to upload all
changes between branch `master` and branch `develop`:

	$ git checkout develop         # This is the version which is uploaded.
	$ git ftp push --commit master # Upload changes compared to master.

# DEFAULTS

Don't repeat yourself. Setting config defaults for git-ftp in .git/config

	$ git config git-ftp.<(url|user|password|syncroot|cacert|keychain|...)> <value>

Everyone likes examples:

	$ git config git-ftp.user john
	$ git config git-ftp.url ftp.example.com
	$ git config git-ftp.password secr3t
	$ git config git-ftp.syncroot path/dir
	$ git config git-ftp.cacert caCertStore
	$ git config git-ftp.deployedsha1file mySHA1File
	$ git config git-ftp.insecure 1
	$ git config git-ftp.key ~/.ssh/id_rsa
	$ git config git-ftp.keychain user@example.com
	$ git config git-ftp.remote-root htdocs
	$ git config git-ftp.disable-epsv 1
	$ git config git-ftp.no-commit 1

After setting those defaults, push to *john@ftp.example.com* is as simple as

	$ git ftp push

If you run into issues with setting up your password please check this [note](#passwords).

# SCOPES

Need different config defaults per each system or environment? Use the so
called scope feature.

Useful if you use multi environment development. Like a development, testing
and a production environment.

	$ git config git-ftp.<scope>.<(url|user|password|syncroot|cacert)> <value>

So in the case below you would set a testing scope and a production scope.

Here we set the params for the scope "testing"

	$ git config git-ftp.testing.url ftp.testing.com:8080/foobar-path
	$ git config git-ftp.testing.password simp3l

Here we set the params for the scope "production"

	$ git config git-ftp.production.user manager
	$ git config git-ftp.production.url live.example.com
	$ git config git-ftp.production.password n0tThatSimp3l

Pushing to scope *testing* alias *john@ftp.testing.com:8080/foobar-path*
using password *simp3l*

	$ git ftp push -s testing

*Note:* The **SCOPE** feature can be mixed with the **DEFAULTS** feature.
Because we didn't set the user for this scope,
git-ftp uses *john* as user as set before in **DEFAULTS**.

Pushing to scope *production* alias *manager@live.example.com* using
password *n0tThatSimp3l*

	$ git ftp push -s production

*Hint:* If your scope name is identical with your branch name. You can skip the
scope argument, e.g. if your current branch is "production":

	$ git ftp push -s

You can also create scopes using the add-scope action.
All settings can be defined in the URL.
Here we create the *production* scope using add-scope

	$ git ftp add-scope production ftp://manager:n0tThatSimp3l@live.example.com/foobar-path

Deleting scopes is easy using the `remove-scope` action.

	$ git ftp remove-scope production

# IGNORING FILES TO BE SYNCED

Add patterns to `.git-ftp-ignore` and all matching file names will be ignored.
The patterns are interpreted as shell glob patterns since version 1.1.0.
Before version 1.1.0, patterns were interpreted as regular expressions.
Here are some glob pattern examples:

Ignoring everything in a directory named `config`:

	config/*

Ignoring all files having extension `.txt`:

	*.txt

Ignoring a single file called `foobar.txt`:

	foobar.txt

Ignoring Git related files:

	.gitignore
	*/.gitignore      # ignore files in sub directories
	*/.gitkeep
	.git-ftp-ignore
	.git-ftp-include
	.gitlab-ci.yml

# SYNCING UNTRACKED FILES

The `.git-ftp-include` file specifies intentionally untracked files that
Git-ftp should upload.
If you have a file that should always be uploaded, add a line beginning with !
followed by the file's name.
For example, if you have a file called VERSION.txt then add the following line:

	!VERSION.txt

If you have a file that should be uploaded whenever a tracked file changes, add
a line beginning with the untracked file's name followed by a colon and the
tracked file's name.
For example, if you have a CSS file compiled from an SCSS file then add the
following line:

	css/style.css:scss/style.scss

If you have multiple source files, you can add multiple lines for each of them.
Whenever one of the tracked files changes, the upload of the paired untracked
file will be triggered.

	css/style.css:scss/style.scss
	css/style.css:scss/mixins.scss

If a local untracked file is deleted, any change of a paired tracked file will
trigger the deletion of the remote file on the server.

All paths are usually relative to the Git working directory.
When using the `--syncroot` option, paths of tracked files
(right side of the colon) are relative to the set syncroot.
Example:

	# upload "html/style.css" triggered by html/style.scss
	# with syncroot "html"
	html/style.css:style.scss

If your *source* file is outside the syncroot,
prefix it with a / and define a path relative
to the Git working directory. For example:

	# upload "dist/style.css" with syncroot "dist"
	dist/style.css:/src/style.scss

It is also possible to upload whole directories.
For example, if you use a package manager like composer, you can upload all
vendor packages when the file composer.lock changes:

	vendor/:composer.lock

But keep in mind that this will upload all files in the vendor folder, even
those that are on the server already.
And it will not delete files from that directory if local files are deleted.

# MAPPING REMOTE FILE PATHS

The `.git-ftp-map` file allows you to modify the path of a file or
folder to one that is accessible on the remote server.
To do so, add the name of the file, followed by a colon and the new path.
For folders, **the directory separator must be appended** to the local and
remote path.

For example, if you have a folder called `public/css` and you would like it to
be uploaded to `public/assets` instead, then add the following line:

	public/css/:public/assets/

And for a file called `all.js` in the `js` folder to be uploaded to the
`assets` folder:

	js/all.js:assets/all.js

# DOWNLOADING FILES (EXPERIMENTAL)

**WARNING:** It can delete local untracked files that are not listed in your
`.git-ftp-ignore` file.

You can use git-ftp to download from the remote host into your repository.
You will need to install the lftp command line tool for that.

	git ftp download

It uses lftp's mirror command to download all files that are different on the
remote host. You can inspect the changes with git-diff.
But if you have some local commits that have not been uploaded to the remote
host, you may not compare to the right version.
You need to compare the downloaded files to the commit that was uploaded last.
This magic is done automatically by

	git ftp pull

It does the following steps for you:

	git checkout <remote-commit>
	git ftp download
	git add --all
	git commit -m '[git-ftp] remotely untracked modifications'
	git ftp catchup
	git checkout <my-branch>
	git merge <new-remote-commit>

If you want to inspect the downloaded changes before merging them into your
current branch, add the option `--no-commit`.
It will stop during the merge at the end of the pull action.
You can inspect the merge result first and can then decide to continue or
abort.

	git ftp pull --no-commit
	# inspect the result and commit them
	git commit
	# or abort the merge
	git merge --abort

If you abort the merge, the downloaded changes will stay in an unreferenced
commit until the Git garbage collector is run.
The commit id will be printed so that you can tag it or create a new branch.

# HOOKS (EXPERIMENTAL)

**This feature is experimental. The interface may change.**

Git-ftp supports client-side hook scripts during the init and the push action.

`pre-ftp-push` is called just before the upload to the server starts, but after
the changeset of files was generated. It can be bypassed with the --no-verify
option.

The hook is called with four parameters.
The first is the used scope or the host name if no scope is used.
The second parameter is the destination URL.
The third is the local commit id which is going to be uploaded and
the fourth is the remote commit id on the server which is going to be updated.

The standard input is a list of all filenames to sync. Each file is preceeded
by A or D followed by a space. A means that this file is scheduled for upload,
D means it's scheduled for deletion. All entries are separated by the NUL byte.
This list is different to git diff, because
it has been changed by the rules of the `.git-ftp-include` file and the
`.git-ftp-ignore` file.

Exiting with non-zero status from this script causes
Git-ftp to abort and exit with status 9.

An example script is:

```bash
#!/bin/bash
#
# An example hook script to verify what is about to be uploaded.
#
# Called by "git ftp push" after it has checked the remote status, but before
# anything has been pushed. If this script exits with a non-zero status nothing
# will be pushed.
#
# This hook is called with the following parameters:
#
# $1 -- Scope name if set or host name of the remote
# $2 -- URL to which the upload is being done
# $3 -- Local commit id which is being uploaded
# $4 -- Remote commit id which is on the server
#
# Information about the files which are being uploaded or deleted is supplied
# as NUL separated entries to the standard input in the form:
#
#   <status> <path>
#
# The status is either A for upload or D for delete. The path contains the
# path to the local file. It contains the syncroot if set.
#
# This sample shows how to prevent upload of files containing the word TODO.

remote="$1"
url="$2"
local_sha="$3"
remote_sha="$4"

while read -r -d '' status file
do
	if [ "$status" = "A" ]
	then
		if grep 'TODO' "$file"; then
			echo "TODO found in file $file, not uploading."
			exit 1
		fi
	fi
done

exit 0
```

`post-ftp-push` is called after the transfer has been finished. The standard
input is empty, but the parameters are the same as given to the `pre-ftp-push`
hook. This hook is **not** bypassed by the --no-verify option, but rather can be
bypassed by the --no-post-hooks. It is meant primarily for notification and its
exit status does not have any effect.

# PASSWORDS

If your password contains special characters you have to take it with care. In most cases it is a good idea to quote passwords with single quotes:

	--passwd '#my$fancy!secret'

Mostly `--ask-passwd` works even if `--passwd` does not work. So maybe you can give this a try.

If your password starts with a hyphen/dash (`-`) even quoting might fail.
This is [by design](https://github.com/git-ftp/git-ftp/issues/468) and will not be fixed.
In this case you can use one of the other options to set your password: the defaults feature using `git config`, `--ask-passwd` or `~/.netrc`.

Quoting also works if a [default](#defaults) is set with `git config`:

	$ git config git-ftp.password '#my$fancy!secret'

## NETRC

In the backend, Git-ftp uses curl.
This means `~/.netrc` could be used beside the other options of Git-ftp
to authenticate.

	$ editor ~/.netrc
	machine ftp.example.com
	login john
	password SECRET

With git-ftp the credentials stored in this file are used if no username is set.
For example, if you set up your .netrc file like this you can just call

	git ftp init ftp.example.com

Of course this can be combined with the [defaults feature](#defaults) to set config defaults for other options as well.

## Password managers

You can use an arbitrary `bash` command to retrieve the password.
In the following examples, we'll use the password manager [pass](https://www.passwordstore.org/).

You can specify the command with the option `--password-command`:

	$ git ftp init --password-command "pass show host"

Or you can set a config for this, so you don’t need to repeat yourself:

	$ git config git-ftp.password-command "pass show host"

Since it is an arbitrary `bash` command, you can for example use pipes:

	$ git ftp push --password-command "pass show host |head -1"

The command-line option takes precedence over the configured value.

## Keychain on macOS

On macOS you can use the built in keychain to store and get your passwords.

You can use this feature by using the option `--keychain` in your command:

	$ git ftp init --keychain account@host ftpes://host

You can omit the value for this option. Then git-ftp will guess the account and hostname from user and url.

Or you can set a config for this, so you don’t need to repeat yourself (see [defaults](#defaults) for details):

	$ git config git-ftp.keychain account@host

You can omit the hostname here. If there is no `@` in the config value git-ftp will guess the hostname from url.


changelog
Version 0.7.3
=============

* Added add-scope and remove-scope actions.
* .git-ftp-ignore can now contain comments (#...) and whitespaces.
* Fixed bug if path to git project contains whitespaces.
* Fixed bug in syncroot feature.
* Removed parallel connections feature.
* Code cleanup (syncroot).


git-ftp.sh
#!/bin/sh
#
# Copyright (c) 2010 
# René Moser <mail@renemoser.net>
# Eric Greve <ericgreve@gmail.com>
# Timo Besenreuther <timo@ezdesign.de>
#

# ------------------------------------------------------------
# Setup Environment
# ------------------------------------------------------------

# General config
DEFAULT_PROTOCOL="ftp"
DEPLOYED_SHA1_FILE=".git-ftp.log"
LCK_FILE="`basename $0`.lck"
REMOTE_LCK_FILE="`basename $0`.lck"

# ------------------------------------------------------------
# Defaults
# ------------------------------------------------------------
URL=""
REMOTE_PROTOCOL=""
REMOTE_HOST=""
REMOTE_USER=${USER}
REMOTE_PASSWD=""
REMOTE_PATH=""
VERBOSE=0
IGNORE_DEPLOYED=0
DRY_RUN=0
FORCE=0
ACTION=""
ENABLE_REMOTE_LCK=0

VERSION='0.0.8'
AUTHORS='Rene Moser <mail@renemoser.net>'
CONTRIBUTORS='Eric Greve <ericgreve@gmail.com>, Timo Besenreuther <timo@ezdesign.de>'
 
usage_long()
{
less << EOF
USAGE
        git-ftp <action> [<options>] <url> [<options>]


DESCRIPTION
        git-ftp does FTP the GIT way. 
        
        It uses GIT to find out which files are different of the uploaded files 
        and let you save time and bandwith by uploading only those files.
        
        It remembers the deployed state by uploading the SHA1 in a log file.
        

ACTIONS
        . push
                Uploads git tracked files which have changed since last upload.
                
        . show
                Gets last uploaded SHA1 from log and hooks git show.
                
        . catchup
                Uploads the current SHA1 to the log, does not upload any files. 
                
                This is useful if you used another FTP client to upload the 
                files and now want to remember the SHA1.
        . help
                Shows this help


URL
        . FTP (default) host.example.com[:<port>][/<remote path>]
        . FTP           ftp://host.example.com[:<port>][/<remote path>]
        . SFTP          sftp://host.example.com[:<port>][/<remote path>]
        . FTPS          ftps://host.example.com[:<port>][/<remote path>]


OPTIONS
        -h, --help      Show this help
        -u, --user      FTP login name
        -p, --passwd    FTP password
        -D, --dry-run   Dry run: Does not upload anything
        -a, --all       Uploads all files, ignores deployed SHA1 hash
        -l, --lock      Enable/Disable remote locking
        -f, --force     Force, does not ask questions
        -v, --verbose   Verbose
        --version       Prints version
        

EXAMPLES
        .   git-ftp push -u john ftp://ftp.example.com:4445/public_ftp -p -v
        .   git-ftp push -p -u john -v ftp.example.com:4445:/public_ftp 


VERSION
        $VERSION


AUTHORS
        $AUTHORS


CONTRIBUTORS
        $CONTRIBUTORS
EOF
exit 0
}

usage() {
    echo "git-ftp <action> [<options>] <url> [<options>]"
    exit 1
}

ask_for_passwd() {
    echo -n "Password: "
    stty -echo
    read REMOTE_PASSWD
    stty echo
    echo ""
}

# Checks if last comand was successful
check_exit_status() {
    if [ $? -ne 0 ]; then
        write_error "${1}, exiting..."
        release_lock
        exit 1
    fi
}

# Simple log func
write_log() {
    if [ $VERBOSE -eq 1 ]; then
        echo "`date`: ${1}"
    fi
}

# Simple error writer
write_error() {
    if [ $VERBOSE -eq 0 ]; then
        echo "fatal: ${1}"
    else
        write_log "fatal: ${1}"
    fi
}

# Simple info writer
write_info() {
    if [ $VERBOSE -eq 0 ]; then
        echo "${1}"
    else
        write_log "${1}"
    fi
}

upload_file() {
    source_file="${1}"
    dest_file="${2}"
    if [ -z ${dest_file} ]; then
        dest_file=${source_file}
    fi
    curl -s -T ${source_file} --user ${REMOTE_USER}:${REMOTE_PASSWD} --ftp-create-dirs -# ${REMOTE_PROTOCOL}://${REMOTE_HOST}/${REMOTE_PATH}${dest_file}
}

remove_file() {
    file=${1}
    curl -s --user ${REMOTE_USER}:${REMOTE_PASSWD} -Q "-DELE ${REMOTE_PATH}${file}" ${REMOTE_PROTOCOL}://${REMOTE_HOST} > /dev/null 2>&1
}

get_file_content() {
    source_file=${1}
    curl -s --user ${REMOTE_USER}:${REMOTE_PASSWD} ${REMOTE_PROTOCOL}://${REMOTE_HOST}/${REMOTE_PATH}${source_file}
}

upload_local_sha1() {
    DEPLOYED_SHA1=`git log -n 1 --pretty=format:%H`
    write_log "Uploading commit log to ${REMOTE_PROTOCOL}://${REMOTE_HOST}/${REMOTE_PATH}${DEPLOYED_SHA1_FILE}"
    if [ ${DRY_RUN} -ne 1 ]; then
        echo "${DEPLOYED_SHA1}" | upload_file - ${DEPLOYED_SHA1_FILE}
        check_exit_status "Could not upload"
    fi
    write_info "Last deployment changed to ${DEPLOYED_SHA1}";
}

check_remote_lock() {
    if [ ${FORCE} -ne 0 ]; then
        return;
    fi

    write_log "Checking remote lock"
    lck_content="`get_file_content ${REMOTE_LCK_FILE} 2>/dev/null`"
    if [ "${lck_content}" != "" ]; then
        lck_sha1=`echo "${lck_content}" | head -n 1`
        write_log "Remote lock sha1 ${lck_sha1}"
        local_sha1=`git log -n 1 --pretty=format:%H`
        write_log "Local sha1 ${local_sha1}"
        if [ "${lck_sha1}" != "${local_sha1}" ]; then
            lck_user=`echo "${lck_content}" | tail -n 1`
            write_error "Remote locked by ${lck_user}"
            release_lock
            exit 1
        fi
    fi
}

remote_lock() {
    if [ ${ENABLE_REMOTE_LCK} -ne 1 ]; then
        return;
    fi

    check_remote_lock

    local_sha1="`git log -n 1 --pretty=format:%H`"
    local_hostname="`hostname --fqdn`"
    date_utc="`date --utc --rfc-2822`"
    lck_message="${USER}@${local_hostname} on ${date_utc}"

    write_log "Remote locking ${lck_message}"
    if [ ${DRY_RUN} -ne 1 ]; then
        echo "${local_sha1}\n${lck_message}" | upload_file - ${REMOTE_LCK_FILE}
        check_exit_status "Could not upload remote lock file"
    fi
}

release_remote_lock() {
    if [ ${ENABLE_REMOTE_LCK} != 1 ]; then
        return;
    fi
    write_log "Releasing remote lock"
    remove_file ${REMOTE_LCK_FILE}
}

# Release lock func
release_lock() {
    write_log "Releasing lock"
    rm -f "${LCK_FILE}"
}

# Handle action
case "${1}" in
    push)
        ACTION="${1}"
        ;;
    catchup)
        ACTION="${1}"
        ;;
    show)
        ACTION="${1}"
        ;;
    help)
        usage_long
        ;;
esac

while test $# != 0
do
	case "${1}" in
	    -h|--h|--he|--hel|--help)
		    usage_long
		    ;;
        -u|--user*)
            case "$#,${1}" in
                *,*=*)
                    REMOTE_USER=`expr "z${1}" : 'z-[^=]*=\(.*\)'`
                    ;;
                1,*)
                    usage 
                    ;;
                *)
                    if [ ! `echo "${2}" | egrep '^-' | wc -l` -eq 1 ]; then
                        REMOTE_USER="${2}"
                        shift                        
                    fi
                    ;;                      
            esac
            ;;
        -p|--passwd*)
            case "$#,$1" in
                *,*=*)
                    REMOTE_PASSWD=`expr "z${1}" : 'z-[^=]*=\(.*\)'`
                    ;;
                1,*)
                    ask_for_passwd 
                    ;;
                *)
                    if [ ! `echo "${2}" | egrep '^-' | wc -l` -eq 1 ]; then
                        REMOTE_PASSWD="${2}"
                        shift
                    else 
                        ask_for_passwd
                    fi
                    ;;
            esac
            ;;
        -a|--all)
            IGNORE_DEPLOYED=1
            ;;
        -l|--lock)
            if [ ${ENABLE_REMOTE_LCK} -ne 1 ]; then
                write_log "Enabling remote locking feature"
                ENABLE_REMOTE_LCK=1
            else
                write_log "Disabling remote locking feature"
                ENABLE_REMOTE_LCK=0
            fi
            ;;
        -D|--dry-run)
            DRY_RUN=1
            write_log "Running dry, won't do anything"            
            ;;
        -v|--verbose)
            VERBOSE=1
            ;;
        -f|--force)
            FORCE=1
            write_log "Forced mode enabled"
            ;;
        --version)
            echo "${VERSION}"
            exit 0
            ;;
        *)
            # Pass thru anything that may be meant for fetch.
            URL=${1}
            ;;
    esac
    shift
done

if [ "${ACTION}" = "" ]; then
    usage
    exit 1
fi

# Checks locking, make sure this only run once a time
if [ -f "${LCK_FILE}" ]; then

    # The file exists so read the PID to see if it is still running
    MYPID=`head -n 1 "${LCK_FILE}"`

    TEST_RUNNING=`ps -p ${MYPID} | grep ${MYPID}`

    if [ -z "${TEST_RUNNING}" ]; then
        # The process is not running echo current PID into lock file
        write_log "Not running"
        echo $$ > "${LCK_FILE}"
    else
        write_log "`basename ${0}` is already running [${MYPID}]"
        exit 0
    fi
else
    write_log "Not running"
    echo $$ > "${LCK_FILE}"
fi

# ------------------------------------------------------------
# Main part
# ------------------------------------------------------------

# Check if this is a git project here
if [ ! -d ".git" ]; then
    write_error "Not a git project? Exiting..."
    release_lock
    exit 1
fi 

# Check if the git working dir is dirty, except for show action
if [ "${ACTION}" != "show" ]; then
    CLEAN_REPO=`git status -uno | egrep "nothing to commit*" | wc -l` 
    if [ ${CLEAN_REPO} -ne 1 ]; then 
        write_error "Dirty Repo? Exiting..."
        release_lock
        exit 1
    fi
fi

# Split host from url
REMOTE_HOST=`expr "${URL}" : ".*://\([a-z0-9\.:-]*\).*"`
if [ -z ${REMOTE_HOST} ]; then
    REMOTE_HOST=`expr "${URL}" : "^\([a-z0-9\.:-]*\).*"`
fi

# Some error checks
HAS_ERROR=0
if [ -z ${REMOTE_HOST} ]; then
    write_error "Remote host not set"
    HAS_ERROR=1
fi

if [ -z ${REMOTE_USER} ]; then
    write_error "Remote user not set"
    HAS_ERROR=1
fi

if [ ${HAS_ERROR} -ne 0 ]; then
    usage
    release_lock
    exit 1
fi

# Split protocol from url 
REMOTE_PROTOCOL=`expr "${URL}" : "^\(ftp\|sftp\|ftps\)://.*"`

# Handle supported protocols and remote path
if [ -z ${REMOTE_PROTOCOL} ]; then
    UNKNOWN_PROTOCOL=`expr "${URL}" : "^\(.*:[/]*\).*"`
    if [ -z ${UNKNOWN_PROTOCOL} ]; then
        write_log "Protocol not set, using default protocol ${DEFAULT_PROTOCOL}://"
        REMOTE_PROTOCOL=${DEFAULT_PROTOCOL}
        REMOTE_PATH=`echo "${URL}" | cut -d '/' -f 2-`
    else
        write_error "Protocol unknown '${UNKNOWN_PROTOCOL}'"
        release_lock
        exit 1
    fi
else
    REMOTE_PATH=`echo "${URL}" | cut -d '/' -f 4-`
fi

# Add trailing slash if missing 
if [ ! -z ${REMOTE_PATH} ] && [ `echo "${REMOTE_PATH}" | egrep "*/$" | wc -l` -ne 1 ]; then
    write_log "Added missing trailing / in path"
    REMOTE_PATH="${REMOTE_PATH}/"  
fi

write_log "Host is '${REMOTE_HOST}'"
write_log "User is '${REMOTE_USER}'"
write_log "Path is '${REMOTE_PATH}'"
write_log "Action is '${ACTION}'"

DEPLOYED_SHA1=""
if [ "${ACTION}" = "show" ]; then
    DEPLOYED_SHA1="`get_file_content ${DEPLOYED_SHA1_FILE}`"
    check_exit_status "Could not get uploaded log file"
    git show "${DEPLOYED_SHA1}"
    release_lock
    exit 0
fi

if [ ${FORCE} -ne 1 ]; then
    # Check if are at master branch
    CURRENT_BRANCH="`git branch | grep '*' | cut -d ' ' -f 2`" 
    if [ "${CURRENT_BRANCH}" != "master" ]; then 
        write_info "You are not on master branch."
        echo -n "Are you sure deploying branch '${CURRENT_BRANCH}'? [Y/n] "
        read answer_branch
        if [ "${answer_branch}" = "n" ] || [ "${answer_branch}" = "N" ]; then
            write_info "Aborting..."
            release_lock
            exit 0
        fi
    fi 
fi

if [ "${ACTION}" = "catchup" ]; then
    upload_local_sha1
    release_lock
    exit 0
fi 

DEPLOYED_SHA1=""
if [ ${IGNORE_DEPLOYED} -ne 1 ]; then
    # Get the last commit (SHA) we deployed if not ignored or not found
    write_log "Retrieving last commit from ${REMOTE_PROTOCOL}://${REMOTE_HOST}/${REMOTE_PATH}"
    DEPLOYED_SHA1="`get_file_content ${DEPLOYED_SHA1_FILE}`"
    if [ $? -ne 0 ]; then
        write_info "Could not get last commit or it does not exist"
        DEPLOYED_SHA1=""
    fi
fi

if [ "${DEPLOYED_SHA1}" != "" ]; then
    write_log "Last deployed SHA1 for ${REMOTE_HOST}/${REMOTE_PATH} is ${DEPLOYED_SHA1}"

    # Get the files changed since then
    FILES_CHANGED="`git diff --name-only ${DEPLOYED_SHA1} 2>/dev/null`" 
    if [ $? -ne 0 ]; then
        if [ ${FORCE} -ne 1 ]; then
        write_info "Unknown SHA1 object, make sure you are deploying the right branch and it is up-to-date."
        echo -n "Do you want to ignore and upload all files again? [y/N] "
        read answer_state
        if [ "${answer_state}" != "y" ] && [ "${answer_state}" != "Y" ]; then
            write_info "Aborting..."
            release_lock
            exit 0
        else
            write_log "Taking all files";
            FILES_CHANGED="`git ls-files`"
        fi
        else 
            write_info "Unknown SHA1 object, could not determine changed filed, taking all files"
            FILES_CHANGED="`git ls-files`"
        fi    
    elif [ "${FILES_CHANGED}" != "" ]; then 
        write_log "Having changed files";
    else 
        write_info "No changed files for ${REMOTE_HOST}/${REMOTE_PATH}. Everything up-to-date."
        release_lock
        exit 0
    fi
else 
    write_log "No last deployed SHA1 for ${REMOTE_HOST}/${REMOTE_PATH} found or forced to take all files"
    FILES_CHANGED="`git ls-files`"
fi

# Calculate total file count
done_items=0
total_items=`echo "${FILES_CHANGED}" | wc -l`
total_items=$((total_items+0)) # trims whitespaces produced by wc
write_log "There are ${total_items} changed files"

# Uploack remote lock file if enabled
remote_lock

# Changing internal field separator, file names could have spaces
OIFS="${IFS}"
NIFS=`echo '\n$'`
IFS="${NIFS}"

# Upload to ftp
for file in ${FILES_CHANGED}; do
    done_items=$(($done_items+1))
    # File exits?
    if [ -f ${file} ]; then 
        # Uploading file
        write_info "[${done_items} of ${total_items}] Uploading '${file}'"
        if [ ${DRY_RUN} -ne 1 ]; then
            upload_file ${file}
            check_exit_status "Could not upload"
        fi
    else
        # Removing file
        write_info "[${done_items} of ${total_items}] Removing not existing file '${file}'"
        if [ ${DRY_RUN} -ne 1 ]; then
            remove_file ${file}
        fi
    fi
done

IFS="${OIFS}"
 
# if successful, remember the SHA1 of last commit
upload_local_sha1

# ------------------------------------------------------------
# Cleanup
# ------------------------------------------------------------
release_remote_lock
release_lock

# ------------------------------------------------------------
# Done
# ------------------------------------------------------------
exit 0
